### DISCUSSION TITLE
Server Pack Exports

### DISCUSSION THREAD

**Discussion by Griefed** [Griefed/ServerPackCreator#345] [Ideas]
Moved from https://github.com/Griefed/ServerPackCreator/issues/324

From @Kreezxil

> Whats the feature you would like to be added?
> 
> Have SPC create a mod manifest, much in the same way that the official pack launchers make their exports. This would allow for a few things, the main part of which will be a drastically smaller server pack (pre-installed) size to upload to CurseForge.
> Anything else you would like to add?
> 
> Currently, most of the FTB mod packs, under the server files section, has a binary that is distributed that acts like a client, it says it's gathering the files and eventually shows itself downloading them.
> 
> However, and I don't remember where, this was not a binary, it was a script that did it.
> 
> I think you're already scanning the cf API anyway, meaning you're probably not that far off from creating this already.
> 
> It would be useful for the super mega packs, which sadly will die off because people would rather not play mod packs that go above a gigabyte anyway. Not until the poverty class computer is a 32 GB standard computer, at least.

> for reference: https://www.feed-the-beast.com/modpack/91_ftb_oceanblock/server-files

From @Griefed 

> Heya Kreezxil,
>
> thanks for this feature request. 🙂
> 
> I have indeed already thought about adding such a feature. Namely, adding an option to export a [ServerStarter](https://github.com/BloodyMods/ServerStarter) compatible format.
> 
>     However, and I don't remember where, this was not a binary, it was a script that did it.
> 
> Correct me if I am wrong. This sounds a lot like: https://github.com/BloodyMods/ServerStarter
> I have no idea how the FTB people generate the OS binaries which you linked below, but I would prefer OS-independant exports anyway. Generate once, run anywhere. Plus, if their installers download from CurseForge, we run into the CurseForge API and token problem again. See below.
> 
>     I think you're already scanning the cf API anyway, meaning you're probably not that far off from creating this already.
> 
> Unfortunately, due to the changes/limitations to the CurseForge API, I am not. At least not currently.
> A rewrite of the way SeverPackCreator accesses CurseForges API is still planned, but it will require the user to set a valid CurseForge API token in the configuration in order to access CurseForge related features. Otherwise it will create either of the following situations:
> 
>     Some checks/values are gathered from CurseForge until a rate limit is hit, thus effectively killing the feature for any potential user and generating incomplete server packs, i.e garbage.
>     A given user, without a valid CurseForge API token, can not generate such packages/exports etc..
> 
> It is not possible to ship a CurseForge API token with ServerPackCreator without compromising it. ServerPackCreator is open source, so anyone can figure out how to get a hold of a token, should any be embedded into it. Or they could use a decompiler and get it this way.
> 
> If and when the API rewrite for SPC is done, I may consider this as a feature for the webservice. There, the admin(s) who run it can set a token and any generated server pack could have the option to additionally include a ServerStarter export.
> 
> Of course, tests need to be made whether ServerStarter still works after the CurseForge API changes. It may be entirely possible that ServerStarter has become useless due to these changes. So, preferrably, ServerStarter tests must happen before I start working on this. If ServerStarter has become unreliable due to CurseForges changes, this feature is pretty much already dead, unfortunately.
> 
> I don't see any other feasible way of reliably implementing this.

> TL;DR: Exports would be awesome, but thanks to CurseForges API changes, they would only be available on a webservice instance with a valid token configured. ServerStarter functionality must be evaluated first, then I can come back to this when the CurseForge API implementation is done.

From @Griefed 

> CurseForge/Modrinth API implementation has been scrapped completely.
> 
> I have yet to come up with a good idea how server pack exports could be realized in a useful way and whether this feature would make sense for ServerPackCreator. Personally, I would put this under "Out of scope", but oh well.
> 
> Modrinth manifests provide a lot of information about a modpack, and ServerPackCreator is going to make use of them. Maybe someone will come up with a good idea sometime, but currently, that someone is not me. 😂

From @Kreezxil

> No pressure, just leave it on the back burner, close it for now. but maybe
pin it too, so someone can see it's been requested. That way someone
willing to figure out the alchemy required and which type of goat the
demons require for a sacrifice can then make that part happen. Maybe add a
tag to it such as "good idea for a plugin".

Use this discussion to talk about possible ways of creating a server pack export feature which does not involve CurseForges or Modrinths API.
Additionally, individual mods **must not** be hosted by any service created through this export-feature. Most mod-devs, rightfully so, do not allow re-hosting.

**Comment by Griefed** [comment]
Can be done for UNIX/Linux using https://makeself.io/, otherwise, on Windows, it's possible to create self-extracting archives. People may use these tools to create one-file-distrubtions. ServerPackCreator generates the server pack, and if people want, an archive for upload on CurseForge or other platforms.

**Reply by kreezxil** [reply]
I'm sure I wasn't referring to a self-extracting archive. Which was all the rage during my Amiga OS days.

What I was referring to was basically that spc creates it's own manifest to use for downloading mods and other content from cf and modrinth and other sources as required. Meaning it wouldn't even come with mods anymore unless they can't be downloaded.

Still, this is just a "neat if you can" feature, not a "would be awesome if it did" feature. Granted, I could see it lessing the load on hosting sites as they wouldn't have to hold all those jars in the zips if the server packs were download content instead of holding content.

**Reply by Griefed** [reply]
The most I can offer is the solution addressed for Modrinth for the webservice mentioned here: https://github.com/Griefed/ServerPackCreator/issues/324

**Comment by kreezxil** [comment]
honestly, although I posited it. It's not a requirement. But if anyone else wants it they should take note of what you're saying about #324
